### PR TITLE
avoid publishing if nothing changes - initial version, with comparing only poms  #84

### PR DESCRIPTION
**Pull request by wwilk**
@szczepiq 

Like I said on Slack this is a draft, just to show you the dependencies that I made between plugins. 
Right now it runs like that: 
:bumpVersionFile
:fetchNotableReleaseNotes
...
:gitCommit
:gitPush
...
:{subprojects}:jar
...
:{subprojects}:publishToMavenLocal
:{subprojects}:comparePublications
:assertReleaseNeeded
:{subprojects}:bintrayUpload
:bintrayUploadAll
:performRelease

If all of the publications are equal it fails on :assertReleaseNeeded

If I didn't miss anything there is no configuration or runtime dependencies between two hierarchies of plugins. BaseJavaLibraryPlugin checks if bumpVersionFile task is available before using it, so if there is no VersioningPlugin, release will be always needed because there is no way to compare publications without versions. 

ContinuousDeliveryPlugin only makes assertReleaseNeeded dependent on comparePublications if there are projects who use BaseJavaLibraryPlugin.

### CODE REVIEW

File: src/test/groovy/org/mockito/release/version/DefaultVersionInfoTest.groovy [hunk 1]
```diff
@@ -112,7 +115,45 @@ notableVersions=1.0.0
         then:
         f.text == """version=1.0.1
 notableVersions=1.0.0
+previousVersion=1.0.0
 """
         v.notableVersions.toString() == ["1.0.0"].toString()
     }
+
+    def "bumps previous version"() {
```
Comment (szczepiq) [inline src/test/groovy/org/mockito/release/version/DefaultVersionInfoTest.groovy]: Great unit test. Makes it really clear to understand the new behavior.

**Comment by szczepiq** [comment]
Nice work! Really looking forward to enabling this logic in Mockito project.

I also very much like the previous version setting in the version file and the version info object. We will leverage it for release notes!

1. What do you think about a separate plugin “PublicationsComparatorPlugin”, that would apply “BaseJavaLibraryPlugin” and “VersioningPlugin". “JavaLibraryPlugin” would apply “PublicationsComparatorPlugin” and "“BaseJavaLibraryPlugin”. This way, the “BaseJavaLibraryPlugin” remains simple, decoupled from the publication comparison logic and reusable outside of releasing. We should start documenting the diagrams of plugins :)

2. I’m not sure why "compare publications task” depends on bumpVersion file. It would be kind of awkward because each time you compare publications, the version will be incremented. Here’s the workflow I would suggest:

a) assertReleaseNeeded depends on new task comparePublicationsAll
b) comparePublicationsAll depends on comparePublications in all submodules (similar to how we do it with bintrayUploadAll). comparePublicationsAll also keeps all tasks it depends on in a list / some field on the task type. When executed, comparePublicationsAll will iterate over the comparePublications tasks and see if all are equal
b) assertReleaseNeeded is intended to be called as a separate Gradle invocation. In travis.yml of the example project we do:

```
after_success:  - ./gradlew assertReleaseNeeded && ./gradlew travisReleasePrepare && ./gradlew performRelease
```

This is the simplest way to implement ‘release needed functionality’. If the assert task fails, other commands (travisReleasePrepare, performRelease) will not be executed at all. Note that we don’t need to run ‘bumpVersion’ task at all in above workflow. We just compare current version (you get it from version.properties) with previous version.

I'll make some tweaks to your code and push to your branch :)

File: src/main/groovy/org/mockito/release/gradle/ReleaseNeededTask.java [hunk 2]
```diff
@@ -29,6 +29,7 @@
 
     private String branch;
     private String releasableBranchRegex;
+    private boolean allPublicationsEqual;
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/gradle/ReleaseNeededTask.java]: Nice new property. Modeling the API like that makes the code and API clean.

File: src/main/groovy/org/mockito/release/gradle/ReleaseNeededTask.java [hunk 3]
```diff
@@ -78,7 +87,8 @@ public void setReleasableBranchRegex(String releasableBranchRegex) {
                 "\n    - skip by env variable: " + skipEnvVariable +
                 "\n    - skip by commit message: " + skippedByCommitMessage +
                 "\n    - is pull request build:  " + pullRequest +
-                "\n    - is releasable branch:  " + releasableBranch;
+                "\n    - is releasable branch:  " + releasableBranch +
+                "\n    - anything changed in publications since the last release:  " + allPublicationsEqual;
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/gradle/ReleaseNeededTask.java]: This is also nice :) Will make it really helpful when debugging stuff 🐛 :)

File: src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java [hunk 4]
```diff
@@ -26,17 +34,142 @@ public boolean isPublicationsEqual() {
...
+        FileOutputStream output = null;
+        try {
+            String url = getRemoteUrl(extension);
+            input = new BufferedInputStream(new URL(url).openStream());
+            String localUrl = getLocalArtifactUrl(getPreviousVersion(), extension);
+
+            IOUtil.createParentDirectory(new File(localUrl));
+
+
+            FileOutputStream fos = new FileOutputStream(localUrl);
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java]: Maybe you can move lines 89-109 to `IOUtil`?

File: src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java [hunk 5]
```diff
@@ -26,17 +34,142 @@ public boolean isPublicationsEqual() {
...
+
+
+    /**
+     *
+     * @param extension, suffix of artifact eg ".pom" or "-sources.jar"
+     * @return eg
+     * https://bintray.com/shipkit/examples/download_file?file_path=/org/mockito/release-tools-example/api/0.15.1/api-0.15.1.pom";
+     */
+    private String getRemoteUrl(String extension){
+        return "https://bintray.com/shipkit/examples/download_file?file_path="
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/comparison/PublicationsComparatorTask.java]: Please don't forget that not everyone use bintray ;-)

**Comment by wwilk** [comment]
@mstachniuk 
Thanks for the review! As I said above, it's a draft, that's why Bintray is hardcoded, I wanted to connect everything together first and then focus on details. I wrote about that in corresponding issue - the idea is to allow user to provide us with a closure that will return download URL based on some parameters (group, name, version etc). This way he can use whatever remote repository he wants. 

Good idea with moving download code to IOUtil, I'll do that!

**Comment by szczepiq** [comment]
Pushed some more Gradle API guidance. Hope that helps!

**Comment by wwilk** [comment]
@szczepiq Thanks! I committed a few things, but still not done, will be tomorrow probably.

File: src/main/groovy/org/mockito/release/internal/comparison/BaseProjectProperties.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/comparison/BaseProjectProperties.java]: Don't forget to run checkstyle before merge

File: src/main/groovy/org/mockito/release/internal/comparison/ZipComparator.java [hunk 7]
```diff
@@ -1,52 +1,133 @@
...
 
-    Result compareFiles() {
-        final File file1 = this.file1.call();
-        final File file2 = this.file2.call();
-        notNull(file1, "zip/jar file to compare", file2, "zip/jar file to compare");
+    boolean compareZips(String filePath1, String filePath2) {
+        ZipFile file1 = openZipFile(filePath1);
+        ZipFile file2 = openZipFile(filePath2);
+
+        Set<String> set1 = new LinkedHashSet<String>();
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/comparison/ZipComparator.java]: Extract method

File: src/main/groovy/org/mockito/release/internal/comparison/ZipComparator.java [hunk 8]
```diff
@@ -1,52 +1,133 @@
...
+        }
+        LOG.info(filecount + " entries matched");
+        if (errcount > 0) {
+            LOG.info(errcount + " entries did not match");
+            return false;
+        }
+        return true;
+    }
+
+    private ZipFile openZipFile(String filePath) {
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/comparison/ZipComparator.java]: Consider create a class ZipUtils and move this method there.

File: src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java [hunk 9]
```diff
@@ -77,6 +86,61 @@ public void execute(Jar jar) {
...
+        TaskMaker.task(project, COMPARE_PUBLICATIONS_TASK, PublicationsComparatorTask.class, new Action<PublicationsComparatorTask>() {
+            public void execute(final PublicationsComparatorTask t) {
+                t.setDescription("Compares artifacts and poms between last version and the currently built one to see if there are any differences");
+
+                project.getRootProject().getPlugins().withType(
+                        VersioningPlugin.class,
+                        new Action<VersioningPlugin>() {
+                            @Override
+                            public void execute(VersioningPlugin versioningPlugin) {
+                                System.out.println("here");
```
Comment (mstachniuk) [inline src/main/groovy/org/mockito/release/internal/gradle/BaseJavaLibraryPlugin.java]: System.out.println()

File: src/test/groovy/org/mockito/release/internal/comparison/PomSiblingDependencyRemoverTest.groovy [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                "</project>"
+
+        when:
+        def result = new PomSiblingDependencyRemover().removeSiblingDependencies(emptyPom,
+                [new BaseProjectProperties("org.mockito", "independent")] as Set)
+
+        then:
+        result == "<project>\n  <dependencies/>\n</project>\n"
+    }
+}
```
Comment (mstachniuk) [inline src/test/groovy/org/mockito/release/internal/comparison/PomSiblingDependencyRemoverTest.groovy]: Consider add <classifier> to tests

Comment (wwilk) [inline src/test/groovy/org/mockito/release/internal/comparison/PomSiblingDependencyRemoverTest.groovy]: What do you mean? That I should cover more cases?

Comment (mstachniuk) [inline src/test/groovy/org/mockito/release/internal/comparison/PomSiblingDependencyRemoverTest.groovy]: Sorry I used `<>` in reply and GitHub cutted a part of message:-)
Some dependencies can use `<classifier>` tag. It can be important for comparator.

Comment (wwilk) [inline src/test/groovy/org/mockito/release/internal/comparison/PomSiblingDependencyRemoverTest.groovy]: Yeah good edge case! I'll add it to a unit test, new version of implementation will handle this situation.

File: src/main/groovy/org/mockito/release/internal/comparison/PomFilter.groovy [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package org.mockito.release.internal.comparison
+
+class PomFilter {
```
Comment (szczepiq) [inline src/main/groovy/org/mockito/release/internal/comparison/PomFilter.groovy]: This class is fantastic! I really like the API and the model, and neat unit test. Thanks!

File: src/test/groovy/org/mockito/release/gradle/ReleaseNeededTaskTest.groovy [hunk 12]
```diff
@@ -0,0 +1,75 @@
+package org.mockito.release.gradle
+
+import org.gradle.api.GradleException
+import org.gradle.testfixtures.ProjectBuilder
+import org.mockito.release.internal.comparison.PublicationsComparator
+import spock.lang.Specification
+
+class ReleaseNeededTaskTest extends Specification {
```
Comment (szczepiq) [inline src/test/groovy/org/mockito/release/gradle/ReleaseNeededTaskTest.groovy]: This is a great test. Thank you very much for adding it!!!!!!!

We were long overdue for the unit test for this task.

**Comment by szczepiq** [comment]
Merging, lets work on outstanding TODOs in separate RBs. For now, I have unhooked the new tasks from the main workflow. So now the code is testable and integrated, but not integrated in workflow.

Looking forward to using this feature in Mockito! ;)
